pandas/core/internals.py

```diff
@@ -4066,6 +4076,27 @@ def canonicalize(block):
         return all(block.equals(oblock)
                    for block, oblock in zip(self_blocks, other_blocks))
 
+    def unstack(self, unstacker):
+        """Return blockmanager with all blocks unstacked"""
+        dummy = unstacker(np.empty((0, 0)), value_columns=self.items)
+        new_columns = dummy.get_new_columns()
+        new_index = dummy.get_new_index()
+        new_blocks = []
+        mask_columns = np.zeros_like(new_columns, dtype=bool)
+
+        for blk in self.blocks:
+            bunstacker = unstacker(
+                blk.values.T, value_columns=self.items[blk.mgr_locs.indexer])
+            new_items = bunstacker.get_new_columns()
+            new_values, mask = bunstacker.get_new_values()
+            new_placement = new_columns.get_indexer(new_items)
+            mask_columns[new_placement] = mask.any(0)
+            new_blocks.extend(blk._unstack(new_values.T, new_placement))
+
+        bm = BlockManager(new_blocks, [new_columns, new_index])
+        bm = bm.take(mask_columns.nonzero()[0], axis=0)
```

Comment:
what is this take for? should be *inside* the loop on bocks (IOW this is on each block)